### PR TITLE
Parser Upgraded

### PR DESCRIPTION
**Pull request by RobertoHE**
TEST 

SETUP:
```
    MIDI2.setHandleError([](int8_t fptr)
                         {
                           Serial.println("\n\nError in parser:" + String(fptr) + "\n\r\n");
                           //delay(1000);
                         });

    MIDI2.setHandleMessage([](const midi::Message<128U> &msg)
                           {
                             Serial.println("\n### Parser ###");
                             Serial.print("Type: ");
                             Serial.print((msg.type), HEX);
                             Serial.println();
                             Serial.println("Lenght: " + String(msg.length));
                             Serial.println("Channel: " + String(msg.channel));
                             if (msg.type == 0xF0)
                             {
                               for (auto i = 0; i < msg.getSysExSize(); i++)
                               {
                                 Serial.print((msg.sysexArray[i]), HEX);
                                 Serial.print(" ");
                               }
                               Serial.println();
                             }
                             else
                             {
                               Serial.print(msg.type + msg.channel - 1, HEX);
                               Serial.print(" ");
                               Serial.print((msg.data1), HEX);
                               Serial.print(" ");
                               Serial.println((msg.data2), HEX);
                             }
                           });

xTaskCreatePinnedToCore([](){
                           for(;;){
                           MIDI2.read();
                           vTaskDelay(1 / portTICK_PERIOD_MS);
                           }
                           },
                          "MIDI-READ",
                          3000,
                          NULL,
                          1,
                          NULL,
                          1);
```

TEST: (the same like for MSVS)

```
Serial.println("SETUP OK");

  Serial.println();
  Serial.println("SysEx with RealTime msg in the middle --------");
  byte sysExAndRealTime[] = {0xB0, 0xF4,             // header + timestamp
                             0xF0,                   // start SysEx
                             0x01, 0x02, 0x03, 0x04, // SysEx data
                             // RealTime message in the middle of a SysEx
                             0xF3,                   // timestampLow
                             0xFA,                   // Realtime msg: Start
                             0x05, 0x06, 0x07, 0x08, // rest of sysex data
                             0xF4,                   // timestampLow
                             0xF7};                  // end of SysEx
  BLEMIDI2.receive(sysExAndRealTime, sizeof(sysExAndRealTime));
  MIDI2.read();
  delay(1000);
  Serial.println();

  Serial.println("SysEx ---------");

  byte sysExPart[] = {0xB0, 0xF4,                                     // header + timestamp
                      0xF0,                                           // start SysEx
                      0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, // sysex data
                      0xF4,                                           // timestampLow
                      0xF7};                                          // end of SysEx
  BLEMIDI2.receive(sysExPart, sizeof(sysExPart));
  MIDI2.read();
  delay(1000);
  Serial.println();

  Serial.println("SysEx part 1");
  byte sysExPart1[] = {0xB0, 0xF4,                                      // header + timestamp
                       0xF0,                                            // start SysEx
                       0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08}; // sysex data
  BLEMIDI2.receive(sysExPart1, sizeof(sysExPart1));
  MIDI2.read();
  delay(1000);
  Serial.println();

  Serial.println("SysEx part 2");
  byte sysExPart2[] = {0xB0,                                     // 1 byte header
                       0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, // sysex data (cont)
                       0xF4,                                     // timestampLow
                       0xF7};                                    // end of SysEx
  BLEMIDI2.receive(sysExPart2, sizeof(sysExPart2));
  MIDI2.read();
  delay(1000);
  Serial.println();

  Serial.println("ble Packet with 1 MIDI messages");
  byte blePacketWithOneMIDIMessage[] = {0xA8, 0xC0,
                                        0x90, 0x3E, 0x3E};
  BLEMIDI2.receive(blePacketWithOneMIDIMessage, sizeof(blePacketWithOneMIDIMessage));
  MIDI2.read();
  delay(1000);
  Serial.println();

  Serial.println("ble Packet with 2 MIDI messages");
  byte blePacketWithTwoMIDIMessage[] = {0xA8, 0xC0,
                                        0x90, 0x3E, 0x3E,
                                        0xC1,
                                        0x91, 0x3E, 0x3E};
  BLEMIDI2.receive(blePacketWithTwoMIDIMessage, sizeof(blePacketWithTwoMIDIMessage));
  MIDI2.read();
  delay(1000);
  Serial.println();

  Serial.println("ble Packet with 3 MIDI messages");
  byte blePacketWithThreeMIDIMessage[] = {0xA8, 0xC0,
                                          0x90, 0x3E, 0x3E,
                                          0xC1,
                                          0xF0,
                                          0x01, 0x02,
                                          0xC2,
                                          0xF7,
                                          0xC3,
                                          0x91, 0x3E, 0x3E};
  BLEMIDI2.receive(blePacketWithThreeMIDIMessage, sizeof(blePacketWithThreeMIDIMessage));
  MIDI2.read();
  delay(1000);
  Serial.println();

  Serial.println("2 MIDI messages with multiple running status");

  byte twoMIDIMessageWithRunningStatus[] = {0xA9, 0xAD,
                                            0xD1, 0x74, //Full Midi 2 bytes(afterTouch)
                                            0x73,       //running
                                            0xAE,       //timeStamp
                                            0x72,       //running after timeStamp
                                            0xAF,       //timeStamp
                                            0x71,       //running after timeStamp
                                            0x70,
                                            0x69,
                                            0x68,
                                            0xB2,             //
                                            0x92, 0x36, 0x70, //Full Midi 3 bytes (noteOn)
                                            0xB3,             //
                                            0x93, 0x37, 0x71,
                                            0x38, 0x72,
                                            0x39, 0x73,
                                            0xB4, //
                                            0x40, 0x74};
  BLEMIDI2.receive(twoMIDIMessageWithRunningStatus, sizeof(twoMIDIMessageWithRunningStatus));
  MIDI2.read();
  delay(1000);
  Serial.println();
  Serial.println("END TEST");

  while (1)
  {
  };
```

RESAULTS RUNNIGSTATUS ENABLE

```
SETUP OK

SysEx with RealTime msg in the middle --------

### Parser ###
Type: FA
Lenght: 1
Channel: 0
F9 0 0

### Parser ###
Type: F0
Lenght: 10
Channel: 0
F0 1 2 3 4 5 6 7 8 F7

SysEx ---------

### Parser ###
Type: F0
Lenght: 10
Channel: 0
F0 1 2 3 4 5 6 7 8 F7

SysEx part 1

SysEx part 2

### Parser ###
Type: F0
Lenght: 17
Channel: 0
F0 1 2 3 4 5 6 7 8 9 A B C D E F F7

ble Packet with 1 MIDI messages

### Parser ###
Type: 90
Lenght: 17
Channel: 1
90 3E 3E

ble Packet with 2 MIDI messages

### Parser ###
Type: 90
Lenght: 17
Channel: 1
90 3E 3E

### Parser ###
Type: 90
Lenght: 17
Channel: 2
91 3E 3E

ble Packet with 3 MIDI messages

### Parser ###
Type: 90
Lenght: 17
Channel: 1
90 3E 3E

### Parser ###
Type: F0
Lenght: 4
Channel: 0
F0 1 2 F7

### Parser ###
Type: 90
Lenght: 4
Channel: 2
91 3E 3E

2 MIDI messages with multiple running status

### Parser ###
Type: D0
Lenght: 4
Channel: 2
D1 74 0

### Parser ###
Type: D0
Lenght: 1
Channel: 2
D1 73 0

### Parser ###
Type: D0
Lenght: 1
Channel: 2
D1 72 0

### Parser ###
Type: D0
Lenght: 1
Channel: 2
D1 71 0

### Parser ###
Type: D0
Lenght: 1
Channel: 2
D1 70 0

### Parser ###
Type: D0
Lenght: 1
Channel: 2
D1 69 0

### Parser ###
Type: D0
Lenght: 1
Channel: 2
D1 68 0

### Parser ###
Type: 90
Lenght: 1
Channel: 3
92 36 70

### Parser ###
Type: 90
Lenght: 1
Channel: 4
93 37 71

### Parser ###
Type: 90
Lenght: 1
Channel: 4
93 38 72

### Parser ###
Type: 90
Lenght: 1
Channel: 4
93 39 73

### Parser ###
Type: 90
Lenght: 1
Channel: 4
93 40 74

END TEST
```

RESAULTS RUNNIGSTATUS DISABLE

```
SETUP OK

SysEx with RealTime msg in the middle --------

### Parser ###
Type: FA
Lenght: 1
Channel: 0
F9 0 0

### Parser ###
Type: F0
Lenght: 10
Channel: 0
F0 1 2 3 4 5 6 7 8 F7

SysEx ---------

### Parser ###
Type: F0
Lenght: 10
Channel: 0
F0 1 2 3 4 5 6 7 8 F7

SysEx part 1

SysEx part 2

### Parser ###
Type: F0
Lenght: 17
Channel: 0
F0 1 2 3 4 5 6 7 8 9 A B C D E F F7

ble Packet with 1 MIDI messages

### Parser ###
Type: 90
Lenght: 17
Channel: 1
90 3E 3E

ble Packet with 2 MIDI messages

### Parser ###
Type: 90
Lenght: 17
Channel: 1
90 3E 3E

### Parser ###
Type: 90
Lenght: 17
Channel: 2
91 3E 3E

ble Packet with 3 MIDI messages

### Parser ###
Type: 90
Lenght: 17
Channel: 1
90 3E 3E

### Parser ###
Type: F0
Lenght: 4
Channel: 0
F0 1 2 F7

### Parser ###
Type: 90
Lenght: 4
Channel: 2
91 3E 3E

2 MIDI messages with multiple running status

### Parser ###
Type: D0
Lenght: 4
Channel: 2
D1 74 0

### Parser ###
Type: D0
Lenght: 4
Channel: 2
D1 73 0

### Parser ###
Type: D0
Lenght: 4
Channel: 2
D1 72 0

### Parser ###
Type: D0
Lenght: 4
Channel: 2
D1 71 0

### Parser ###
Type: D0
Lenght: 4
Channel: 2
D1 70 0

### Parser ###
Type: D0
Lenght: 4
Channel: 2
D1 69 0

### Parser ###
Type: D0
Lenght: 4
Channel: 2
D1 68 0

### Parser ###
Type: 90
Lenght: 4
Channel: 3
92 36 70

### Parser ###
Type: 90
Lenght: 4
Channel: 4
93 37 71

### Parser ###
Type: 90
Lenght: 4
Channel: 4
93 38 72

### Parser ###
Type: 90
Lenght: 4
Channel: 4
93 39 73

### Parser ###
Type: 90
Lenght: 4
Channel: 4
93 40 74

END TEST
```
You can compare with https://comparetext.com/index.html

It only changes the lenght of non-0xF0 messages (not used).

### CODE REVIEW

File: src/BLEMIDI_Transport.h [hunk 1]
```diff
@@ -77,7 +75,7 @@ class BLEMIDI_Transport
 
     void endTransmission()
     {
-        if (mTxBuffer[mTxIndex - 1] == MIDI_NAMESPACE::MidiType::SystemExclusiveEnd)
+        if (mTxBuffer[mTxIndex - 1] == 0xF7)
```
Comment (lathoub) [inline src/BLEMIDI_Transport.h:78]: I'd like to keep the typed midiTypes (`MIDI_NAMESPACE::MidiType::`)

File: src/BLEMIDI_Transport.h [hunk 2]
```diff
@@ -90,7 +88,7 @@ class BLEMIDI_Transport
             {
                 mTxBuffer[mTxIndex - 1] = mTimestampLow; // or generate new ?
             }
-            mTxBuffer[mTxIndex++] = MIDI_NAMESPACE::MidiType::SystemExclusiveEnd;
+            mTxBuffer[mTxIndex++] = 0xF7;
```
Comment (lathoub) [inline src/BLEMIDI_Transport.h:91]: I'd like to keep the typed midiTypes (`MIDI_NAMESPACE::MidiType::`)

File: src/BLEMIDI_Transport.h [hunk 3]
```diff
@@ -229,108 +234,158 @@ class BLEMIDI_Transport
...
-                // Too much data
-                // If not System Common or System Real-Time, send it as running status
                 switch (midiType)
                 {
-                case MIDI_NAMESPACE::MidiType::NoteOff:
-                case MIDI_NAMESPACE::MidiType::NoteOn:
-                case MIDI_NAMESPACE::MidiType::AfterTouchPoly:
-                case MIDI_NAMESPACE::MidiType::ControlChange:
-                case MIDI_NAMESPACE::MidiType::PitchBend:
+                case 0x80:
```
Comment (lathoub) [inline src/BLEMIDI_Transport.h:292]: I'd like to keep the typed midiTypes (`MIDI_NAMESPACE::MidiType::`)

File: src/BLEMIDI_Transport.h [hunk 4]
```diff
@@ -229,108 +234,158 @@ class BLEMIDI_Transport
...
+#ifndef RUNNING_ENABLE
                         mBleClass.add(lastStatus);
+#endif
                         mBleClass.add(buffer[i + 1]);
                         mBleClass.add(buffer[i + 2]);
                     }
                     break;
-                case MIDI_NAMESPACE::MidiType::ProgramChange:
-                case MIDI_NAMESPACE::MidiType::AfterTouchChannel:
+                case 0xC0:
```
Comment (lathoub) [inline src/BLEMIDI_Transport.h:309]: I'd like to keep the typed midiTypes (`MIDI_NAMESPACE::MidiType::`)

File: src/BLEMIDI_Transport.h [hunk 5]
```diff
@@ -229,108 +234,158 @@ class BLEMIDI_Transport
...
+            else
+            {
+#ifndef RUNNING_ENABLE
+                auto midiType = lastStatus & 0xF0;
+                if (sysExContinuation)
+                    midiType = 0xF0;
+
+                switch (midiType)
+                {
+                case 0x80:
```
Comment (lathoub) [inline src/BLEMIDI_Transport.h:342]: use typed `MIDI_NAMESPACE::MidiType::`

File: src/BLEMIDI_Transport.h [hunk 6]
```diff
@@ -229,108 +234,158 @@ class BLEMIDI_Transport
...
+            {
+                previousStatus = lastStatus;
+            }
+
             timestampByte = buffer[rPtr++];
             if (timestampByte >= 80) // is bit 7 set?
             {
-                timestamp = setMidiTimestamp(headerByte, timestampByte); 
-                // what do to with the timestamp?
+                auto timestampLow = 0x7f & timestampByte;
```
Comment (lathoub) [inline src/BLEMIDI_Transport.h:387]: use the setMidiTimestamp in `BLEMIDI_Transport.h` (not avail. in the msvc test app)

**Comment by lathoub** [comment]
Thanks for upgrading the parser @RobertoHE - great job (I added some minor comments)

**Comment by lathoub** [comment]
FYI: I fully upgraded the msvc simulator to use the actual BLE code (so no copy of the parser in msvc)

**Review by RobertoHE — COMMENTED** [review]
@lathoub 
Check it, please

**Comment by lathoub** [comment]
thank you and kudos to @RobertoHE for the hard work to a) identify the issues with the BLE parser b) fixing them

**Pull request merged**